R/schedule.R

```diff
@@ -337,7 +337,26 @@ schedule_describe <- function(.schedule) {
 #' @family schedule functions
 #' @export
 get_timezones <- function(connect) {
-  raw_tz <- connect$GET(unversioned_url("timezones"))
+  res <- NULL
+
+  # If the version is greater than or equal to the target version, or if no
+  # version is available, we try the new path.
+  if (!connect_version_lt(safe_server_version(connect), "2024.09.0")) {
+    res <- connect$GET(v1_url("timezones"), parser = NULL)
+    print(res)
+    if (httr::status_code(res) == 404) {
+      # We will try the old URL.
+      res <- NULL 
+    }
+  }
+  if (is.null(res)) {
+    res <- connect$GET(unversioned_url("timezones"), parser = NULL)
+    print(res)
+  }
+  connect$raise_error(res)
```

Comment:
It feels odd to me that we both check the version and then also later check for a 404 and `NULL` the result. What if instead we did something like:

```
res <- tryCatch(
  connect$GET(v1_url("timezones"), parser = NULL),
  error = function(e) {
    res <- connect$GET(unversioned_url("timezones"), parser = NULL)
  }
)
connect$raise_error(res)
```

Would that work? If no, what would we need to make that work?

Reply:
I don't _think_ this quite works.

(1) We had discussed using cached version information on the server to avoid calling the new endpoint if we _know_ Connect is of an older version than has the new endpoint. When we do that, we still need to be able to fall back to just calling both endpoints if the version is unavailable (`Server.HideVersion`). The timezones API isn't one where the cost of two requests is significant, but it's really simple, so I was using it to experiment with that logic.

(2) `connect$GET` only throws an error if `parser` is not `NULL` (see [here](https://github.com/rstudio/connectapi/blob/7762c7acef13145864518482d636278542782ef5/R/connect.R#L126-L148)). This means that we'd never get to the `error` condition in that `tryCatch`. I think the most concise possible version of this logic that does not check the Connect version is in the thumbnail functions (example [here](https://github.com/rstudio/connectapi/blob/7762c7acef13145864518482d636278542782ef5/R/thumbnail.R#L31-L42)).

Adapting that here would look like:

```r
res <- con$GET(
  url = con$v1("timezones"),
  parser = NULL
)
if (httr::status_code(res) == 404) {
  res <- con$GET(
    unversioned_url("timezones"),
    parser = NULL
  )
}
con$raise_error(res)
```

That doesn't check the Connect version, though, and it'd still be nice to avoid calling v1 URLs if we _know_ they aren't available. But perhaps that's a low enough priority that it's not worth implementing here. Or perhaps there's a better way to do it, e.g. registering implementations based on Connect's version when the client is created, which would be totally outside of the scope of this PR.

Reply:
Yeah, I get that. Since we had discussed this approach, I figured a simpler API was a good moment to experiment with it — but I'll cherry-pick those changes into a separate branch and revert/simplify them here.

I think a `tryCatch()` approach can probably work, if we don't pass in `parser = NULL`, because that's what suppresses raising errors (which happens in `Connect$request()`).